### DISCUSSION TITLE
When should reflect be used?

### DISCUSSION THREAD

**Discussion by leonheess** [lit/lit#2263] [Q&A] [answered]
The [Lit docs](https://lit.dev/docs/components/properties/#reflected-attributes) state:

> Attributes should generally be considered input to the element from its owner, rather than under control of the element itself, so reflecting properties to attributes should be done sparingly.

While the [Custom Element Best Practices](https://developers.google.com/web/fundamentals/web-components/best-practices#aim-to-keep-primitive-data-attributes-and-properties-in-sync,-reflecting-from-property-to-attribute,-and-vice-versa) state:

> Aim to keep primitive data attributes and properties in sync, reflecting from property to attribute, and vice versa.

What advice should be followed? We are encountering this because Vue 3 seems to have problems setting attributes/properties on lit elements if reflect is false.

**Comment by kevinpschaaf** [comment, marked as answer]
The platform patterns around this are a bit inconsistent. However in general the "sparingly" part comes down to not trying to violate a principle of a single source of truth for the attribute. Since attributes can be set from the host, ideally internal state changes (e.g. as a result of UI events) would not change the attribute value, since this could lead the the host being out of sync with the last attribute value it set on the child.  However, for properties that are only used as inputs/configuration for the component, it's generally fine to reflect these since the source of truth is still the host.

For practical purposes sometimes we need to reflect internal state changes to attributes for styling purposes, but [Custom State Pseudo Classes](https://wicg.github.io/custom-state-pseudo-class/) are a new platform feature which allow styling the host tag without affecting the external attributes of the component.

We should try and think about how to add clearer guidance around this to the docs.
